pkg/apis/pipeline/v1beta1/resultref_test.go

```diff
@@ -433,3 +434,343 @@ func TestLooksLikeResultRef(t *testing.T) {
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask",
+					Result:       "sumResult",
+				},
+			},
+		}, {
+			name: "multiple substitution",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(tasks.sumTask1.results.sumResult) and another $(tasks.sumTask2.results.sumResult)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask1",
+					Result:       "sumResult",
+				}, {
+					PipelineTask: "sumTask2",
+					Result:       "sumResult",
+				},
+			},
+		}, {
+			name: "multiple substitution with param",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(params.param) $(tasks.sumTask1.results.sumResult) and another $(tasks.sumTask2.results.sumResult)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask1",
+					Result:       "sumResult",
+				}, {
+					PipelineTask: "sumTask2",
+					Result:       "sumResult",
+				},
+			},
+		}, {
+			name: "first separator typo",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(task.sumTasks.results.sumResult)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: nil,
+		}, {
+			name: "third separator typo",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(tasks.sumTasks.result.sumResult)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: nil,
+		}, {
+			name: "param substitution shouldn't be considered result ref",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "$(params.paramName)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: nil,
+		}, {
+			name: "One bad and good result substitution",
+			args: args{
+				whenExpression: v1beta1.WhenExpression{
+					Input:    "good -> $(tasks.sumTask1.results.sumResult) bad-> $(task.sumTask2.results.sumResult)",
+					Operator: selection.In,
+					Values:   []string{"foo"},
+				},
+			},
+			want: []*v1beta1.ResultRef{
+				{
+					PipelineTask: "sumTask1",
+					Result:       "sumResult",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			expressions, ok := tt.args.whenExpression.GetVarSubstitutionExpressions()
+			if !ok && tt.want != nil {
+				t.Fatalf("expected to find expressions but didn't find any")
+			} else {
+				got := v1beta1.NewResultRefs(expressions)
+				if d := cmp.Diff(tt.want, got); d != "" {
+					t.Errorf("TestNewResultReference/%s %s", tt.name, diff.PrintWantGot(d))
+				}
+			}
+		})
+	}
+}
+
+func TestHasResultReferenceWhenExpression(t *testing.T) {
+	type args struct {
+		whenExpression v1beta1.WhenExpression
+	}
```

Comment:
could you use whenExpression directly instead of wrapping it in the args struct?